document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java

```diff
@@ -546,6 +581,89 @@ private int[] bulkUpsertImpl(Map<Key, Document> documents) throws SQLException,
     }
   }
 
+  private long bulkUpdateRequestsWithFilter(List<BulkUpdateRequest> requests) throws IOException {
+    // Note: We cannot batch statements as the filter clause can be difference for each request. So
+    // we need one PreparedStatement for each request. We try to update the batch on a best-effort
+    // basis. That is, if any update fails, then we still try the other ones.
+    long totalRowsUpdated = 0;
+    Exception sampleException = null;
+
+    for (BulkUpdateRequest request : requests) {
+      Key key = request.getKey();
+      Document document = request.getDocument();
+      Filter filter = request.getFilter();
+
+      try {
+        totalRowsUpdated += update(key, document, filter).getUpdatedCount();
```

Comment:
As a follow-up, we can explore `java.sql.Statement.addBatch` - https://github.com/JetBrains/jdk8u_jdk/blob/master/src/share/classes/java/sql/Statement.java#L520

However, its not on PreparedStatement.